[SPARK-47094][SQL] SPJ : Dynamically rebalance number of buckets when they are not equal

szehon-ho opened the pull request:
### What changes were proposed in this pull request?
-- Allow SPJ between 'compatible' bucket funtions
-- Add a mechanism to define 'reducible' functions, one function whose output can be 'reduced' to another for all inputs.

  ### Why are the changes needed?
-- SPJ currently applies only if the partition transform expressions on both sides are identifical.

  ### Does this PR introduce _any_ user-facing change?
No

  ### How was this patch tested?
Added new tests in KeyGroupedPartitioningSuite

  ### Was this patch authored or co-authored using generative AI tooling?
No

szehon-ho committed:
[SPARK-47094][SQL] SPJ : Dynamically rebalance number of buckets when they are not equal

  ### What changes were proposed in this pull request?
-- Allow SPJ between 'compatible' bucket funtions
-- Add a mechanism to define 'reducible' functions, one function whose output can be 'reduced' to another for all inputs.

  ### Why are the changes needed?
-- SPJ currently applies only if the partition transform expressions on both sides are identifical.

  ### Does this PR introduce _any_ user-facing change?
No

  ### How was this patch tested?
Added new tests in KeyGroupedPartitioningSuite

  ### Was this patch authored or co-authored using generative AI tooling?
No

himadripal commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1537,6 +1537,18 @@ object SQLConf {
       .createWithDefault(false)
 
+  val V2_BUCKETING_ALLOW_COMPATIBLE_TRANSFORMS =
+    buildConf("spark.sql.sources.v2.bucketing.allow.enabled")
```
may be rename `spark.sql.sources.v2.bucketing.allow.enabled` to `spark.sql.sources.v2.bucketing.allowCompatibleTransform.enabled`

szehon-ho replied:
Thanks, yea it is a copy and paste error.

szehon-ho committed:
Fix scalastyle and address review comment

himadripal commented:
was trying this test case, 
`val partition1 = Array(Expressions.years("ts"),
  bucket(2, "id"))
val partition2 = Array(Expressions.days("ts"),
  bucket(4, "id"))`

this throws an exception when paritiallyClustered = false and allowCompatibleTransform=true. 

exception :
`class java.lang.Integer cannot be cast to class java.lang.Long (java.lang.Integer and java.lang.Long are in module java.base of loader 'bootstrap')
java.lang.ClassCastException: class java.lang.Integer cannot be cast to class java.lang.Long (java.lang.Integer and java.lang.Long are in module java.base of loader 'bootstrap')
	at scala.runtime.BoxesRunTime.unboxToLong(BoxesRunTime.java:103)
	at org.apache.spark.sql.catalyst.expressions.BaseGenericInternalRow.getLong(rows.scala:42)`

sunchao commented:
Thanks @szehon-ho ! will take a look in a few days.

szehon-ho committed:
Fix bug of using un-projected expression in sorting

szehon-ho commented:
Thanks @himadripal I was able to modify one of my tests to also reproduce it, and it should be fixed now.

szehon-ho committed:
Add licenses

szehon-ho committed:
Fix compiler warning

advancedxy reviewed (commented):
Hi @szehon-ho,  great to see this feature landing on the Spark side.

To be honest, I am implementing a similar feature called bucket rescaling, but on the Iceberg side. By bucket rescaling, it means that bucketed Iceberg table could report its partitioning with an on-demand bucket number(compatible with current spec's bucket number, a.k.a `on_demand_bucket_number % current_bucket_num = 0 or current_bucket_number % on_demand_bucket_number = 0`). It's also possible for the bucketed iceberg table to have mixed but compatible bucket numbers in its specs.

So, I'd like to share two cents about this topic, hopes that it can share some insights. 
1. besides down scale a bucket, it's also possible to upscale a bucket as long as we put a `bucket_id filter` after the the scan node. For example, to upscale bucket number from 4 to 8 for the bucket_id = 3 bucket. It could be divided into two buckets: `bucket_id_8=3` with filter `bucket(8, id) = 3` and `bucket_id_8 = 7` with filter `bucket(8, id) = 7`
2. it's possible to support bucket number evolution in one bucket partition field.

Anyway, it's game change for Spark's SPJ to support compatible bucket numbers.  The up scale cases might be supported in follow-up prs.

szehon-ho commented on sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala:
```diff
@@ -1310,6 +1314,312 @@ class KeyGroupedPartitioningSuite extends DistributionAndOrderingSuiteBase {
+          val scans = collectScans(df.queryExecution.executedPlan).map(_.inputRDD.
+            partitions.length)
+
+          val expectedBuckets = Math.min(table1buckets, table2buckets)
```
Yes, in fact it is definitely a follow up to do this with 'partiallyClustered' (which is currently not enabled).  This mode today 'duplicate' partitions for the partially clustered side, and I think can be used to turn on the same for compatible transforms.

sunchao replied:
Yes, I think it will be very useful to make partially clustered distribution work with the feature introduced in this PR.

sunchao commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:
```diff
@@ -635,6 +636,22 @@ trait ShuffleSpec {
+   * Returning none also indicates that none of the partition expressions can be reduced on the
+   * corresponding expression on the other shuffle spec.
+   */
+  def reducers(spec: ShuffleSpec): Option[Seq[Option[Reducer[_]]]] = None
```
maybe this should belong to `KeyGroupedShuffleSpec`? I'm not sure whether this API is meaningful for other shuffle specs.

szehon-ho replied:
done

sunchao commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:
```diff
@@ -829,20 +846,59 @@ case class KeyGroupedShuffleSpec(
     }
   }
 
+  override def canCreatePartitioning: Boolean = SQLConf.get.v2BucketingShuffleEnabled &&
```
nit: seems unnecessary change

szehon-ho replied:
done, i think i was trying to move the private method to the bottom

sunchao commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java:
```diff
@@ -0,0 +1,42 @@
+ * @since 4.0.0
+ */
+@Evolving
+public interface ReducibleFunction<T, A> extends ScalarFunction<T> {
```
what is `T` and what is `A`? should `Reducer` also have two type parameters?

szehon-ho replied:
added javadocs

sunchao commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:
```diff
@@ -829,20 +846,59 @@ case class KeyGroupedShuffleSpec(
+    other match {
+      case otherSpec: KeyGroupedShuffleSpec =>
+        val results = partitioning.expressions.zip(otherSpec.partitioning.expressions).map {
+          case (e1: TransformExpression, e2: TransformExpression)
```
maybe it's better to move this logic into `TransformExpression` itself, e.g., add a `TransformExpression.reducer` method.

szehon-ho replied:
done

sunchao commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:
```diff
@@ -635,6 +636,22 @@ trait ShuffleSpec {
+   * Returning none also indicates that none of the partition expressions can be reduced on the
+   * corresponding expression on the other shuffle spec.
+   */
+  def reducers(spec: ShuffleSpec): Option[Seq[Option[Reducer[_]]]] = None
```
One use case I think could be `hours(col)` vs `days(col)`, in this case we can "reduce" the former into the latter.

szehon-ho committed:
Review comments

szehon-ho commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * If this function is 'reducible' on another function, return the {@link Reducer} function.
+     * @param other other function
+     * @param thisArgument argument for this function instance
```
Added javadocs

szehon-ho committed:
Try to fix doc

szehon-ho committed:
Second round review comments

szehon-ho commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java:
```diff
@@ -0,0 +1,72 @@
+     *     <li>other param = Int(2)</li>
+     * </ul>
+     * @param other the other function
+     * @param thisParam param for this function
```
OK, made default impl for both so that implementations can pick one to override

szehon-ho replied:
Actually on second thought, we do plan to add geo partition transforms , for example xz2 ordering https://github.com/spatialx-project/geolake/blob/main/api/src/main/java/org/apache/iceberg/transforms/ExtendedZCurve.java which need resolution.

as well as multi-arg bucket transforms: https://docs.google.com/document/d/1aDoZqRgvDOOUVAGhvKZbp5vFstjsAMY4EFCyjlxpaaw/ 

So i added back the type to Object in latest pr, hope it is ok. 

Note, I tried awhile to make the method type-parameterized like

```
<T> Option<Reducer<I, O>> reducer(ReducibleFunction<I, O> other, T numBuckets,
                                  T otherNumBuckets);
```

 but was not able to override it successfully in Bucket in scala

szehon-ho committed:
Fix linting

szehon-ho committed:
Make reducer take generic arguments

szehon-ho commented:
@sunchao can you take another look?  I guess open question is how to support other potential parameterized transforms in the future, let me know your thoughts.

sunchao commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:
```diff
@@ -846,6 +879,20 @@ case class KeyGroupedShuffleSpec(
 }
 
+object KeyGroupedShuffleSpec {
+  def reducePartitionValue(row: InternalRow,
```
nit: we usually follow this style:

```scala
  def reducePartitionValue(
      row: InternalRow,
      expressions: Seq[Expression],
      reducers: Seq[Option[Reducer[_, _]]]):
    InternalRowComparableWrapper = {
    val partitionVals = row.toSeq(expressions.map(_.dataType))
```

sunchao commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1541,6 +1541,18 @@ object SQLConf {
 
+  val V2_BUCKETING_ALLOW_COMPATIBLE_TRANSFORMS =
+    buildConf("spark.sql.sources.v2.bucketing.allowCompatibleTransforms.enabled")
+      .doc("Whether to allow storage-partition join in the case where the partition transforms" +
```
nit: space after `transforms`

sunchao commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java:
```diff
@@ -0,0 +1,106 @@
+   */
+  default Reducer<I, O> reducer(
+    Object thisParam,
+    ReducibleFunction<?, ?> otherFunction,
```
`ReducibleFunction<I, O> otherFunction` did not work? just wonder why

szehon-ho replied:
Yea I think I get an error like:

```
/Users/szehon/repos/apache-spark/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TransformExpression.scala:105:42: type mismatch;
[error]  found   : org.apache.spark.sql.connector.catalog.functions.ReducibleFunction[_$5,_$6] where type _$6, type _$5
[error]  required: org.apache.spark.sql.connector.catalog.functions.ReducibleFunction[_$3,_$4]
[error]         thisFunction.reducer(numBuckets, otherFunction, otherNumBuckets)
[error]                                          ^
[error] /Users/szehon/repos/apache-spark/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TransformExpression.scala:106:38: type mismatch;
[error]  found   : org.apache.spark.sql.connector.catalog.functions.ReducibleFunction[_$5,_$6] where type _$6, type _$5
```

But actually there is no requirement they have to be the exact same Input, Output type, in theory the other function could have different types.

sunchao reviewed (commented):
Thanks @szehon-ho , looks much better now. I mainly have one more question on the params.

szehon-ho commented on sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java:
```diff
@@ -0,0 +1,106 @@
+   *     <li>otherParam = Int(2)</li>
+   * </ul>
+   *
+   * @param thisParam parameter for this function
```
I was thinking a list, would that work or cause any issue ?

sunchao replied:
I'm thinking whether we can use `InternalRow` for this, which is consistent with the v2 function API (see `ScalarFunction` and `AggregateFunction`, and can support multiple parameters.

We need to define a clear contract between Spark and the callers of this method though: **what should a caller expect to see from `thisParam` and `otherParam`**. 

For `bucket` it is quite clear, but I'm not sure how this will this work with geohash function that you pointed out. In Spark, we need a way for the data source provider to pass the `resolution` of the function to Spark via v2 catalog, and then pass it back to the data source via `reducer`. Any idea?

szehon-ho replied:
Yea I think you are right, it requires more changes to Spark, V2ExpressionUtil, etc to add these.  Let me revert back to bucketReducer() and then later if we figure that out, we can make this just call the general case.

szehon-ho committed:
Review comments

szehon-ho commented:
@sunchao thanks ! addressed review comments.  Lmk if supporting generic function parameter is too messy, and we want to switch back to just bucket for first cut.

szehon-ho committed:
Revert to bucketReducer function

szehon-ho commented:
@sunchao if you have time for another look, reverted to use specific argument type and method for bucket, and worry about other parameterized transforms later.

sunchao reviewed (commented):
Looks mostly good!

szehon-ho committed:
Formatting

szehon-ho committed:
Simplify bucket reducer logic

szehon-ho commented:
@sunchao thanks can you take another look?  I think the failed test is not related (looks like error in pyspark uploading test report).

sunchao reviewed (approved):
LGTM

himadripal reviewed (approved):
Thank you for adding this feature. @szehon-ho

sunchao commented:
Thanks, merged to master!

szehon-ho commented:
Thanks @sunchao and all for the review!

rekbun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:
```diff
@@ -164,6 +165,18 @@ case class BatchScanExec(
           }
 
+          // Also re-group the partitions if we are reducing compatible partition expressions
+          val finalGroupedPartitions = spjParams.reducers match {
```
I believe this could produce incorrect result when joining presorted bucketed tables with compatible bucket counts. 

Specifically, if we have two tables:
1. Bucketed and sorted on the same join keys
2. With different bucket counts, where one table's bucket count is a multiple of the other


When performing a bucketed join in Spark, it's expected that the sort order should be preserved. However, it appears that the merging or grouping process involved in the join might break these sorting guarantees, leading to incorrect results.
